### PR TITLE
Compile with dotty

### PR DESCRIPTION
**Pull request by smarter**
This is all that's needed to make the backend part of the dotty bootstrap, however there are many files in the backend that should not be compiled by dotty at all, in my branch of dotty I ignore them using sbt filters:
```scala
      unmanagedSourceDirectories in Compile ++=
        (((baseDirectory.value / ".." / "scala-backend" / "src" / "compiler" / "scala" / "tools" / "nsc" / "backend") *
          (GlobFilter("*.scala") - "JavaPlatform.scala" - "Platform.scala")) +++
         ((baseDirectory.value / ".." / "scala-backend" / "src" / "compiler" / "scala" / "tools" / "nsc" / "backend" / "icode") **
          (NothingFilter || "Opcodes.scala" || "Primitives.scala")) +++
         ((baseDirectory.value / ".." / "scala-backend" / "src" / "compiler" / "scala" / "tools" / "nsc" / "backend" / "jvm") *
           (GlobFilter("*.scala") - "BCodeICodeCommon.scala"
             - "GenASM.scala" - "ScalacBackendInterface.scala"))
        ).get,
```

An alternative would be to simply completely comment out the content of files that do not compile with dotty, this would make it easier to compile dotty without sbt since you could just pass it `scala-backend/src/compiler/scala/tools/nsc/backend/**/*.scala` instead of having to reproduce the filter.

### CODE REVIEW

**Comment by DarkDimius** [comment]
You are commenting out quite some files from this repo. I'd prefer to keep them so that we could still build a usable-scalac out of it. It used to help me a lot when debugging the difference between Dotty & scalac.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala [hunk 1]
```diff
@@ -23,7 +23,7 @@ import scala.tools.nsc.backend.icode.Primitives.{NE, EQ, TestOp, ArithmeticOp}
  */
 trait BCodeIdiomatic {
   val int: BackendInterface
-  lazy val bTypes = new BTypesFromSymbols[int.type](int)
+  final lazy val bTypes = new BTypesFromSymbols[int.type](int)
```
Comment (DarkDimius) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala:25]: why final?

Comment (smarter) [reply src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala:25]: I got that from @felixmulder I think, do you recall why this makes a difference?

File: src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala [hunk 2]
```diff
@@ -27,7 +27,8 @@ class BTypesFromSymbols[I <: BackendInterface](val int: I) extends BTypes {
   import bCodeAsmCommon._
 
   // Why the proxy, see documentation of class [[CoreBTypes]].
-  val coreBTypes = new CoreBTypesProxy[this.type](this)
+  // Dotty deviation: This type signature should nto be needed.
+  val coreBTypes: CoreBTypesProxy[this.type] = new CoreBTypesProxy[this.type](this)
```
Comment (DarkDimius) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala]: would you mind reporting all deviations?

Comment (smarter) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala]: Sure, I wanted to minimize them first, but maybe that's not needed.

**Comment by smarter** [comment]
> You are commenting out quite some files from this repo. I'd prefer to keep them so that we could still build a usable-scalac out of it. It used to help me a lot when debugging the difference between Dotty & scalac.

This could be done by splitting the companion objects in separate file and not compiling the files with `Global` in dotty, but this would make the diff with scalac larger, I'm fine with doing it that way if you want.

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala (hunk 1 above)
Comment (felixmulder) [reply src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala:25]: Hmm, IIRC this was something fundamental to Dotty that we couldn't get around. I ended up asking @odersky who said it should be final.

Compiling without it will give you an error message - if you can compile it without `final`, that'd jog my memory.

Comment (DarkDimius) [reply src/compiler/scala/tools/nsc/backend/jvm/BCodeIdiomatic.scala:25]: My guess would be https://github.com/lampepfl/dotty/issues/50 & https://github.com/lampepfl/dotty/pull/1051

File: src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala (hunk 2 above)
Comment (smarter) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala]: Done.

**Comment by smarter** [comment]
@DarkDimius I just updated the PR so that it still compiles with scalac (by setting a published version number in version.properties and doing "ant build"). Once this is in I'll make a second PR to fix procedure syntax warnings and this should be all that's needed for the hyper bootstrap.

File: src/compiler/scala/tools/nsc/backend/ScalaPrimitivesOps.scala [hunk 3]
```diff
@@ -0,0 +1,238 @@
+/* NSC -- new Scala compiler
+ * Copyright 2005-2013 LAMP/EPFL
+ * @author  Martin Odersky
+ */
+
+package scala
+package tools.nsc
+package backend
+
+object ScalaPrimitivesOps extends ScalaPrimitivesOps
```
Comment (DarkDimius) [inline src/compiler/scala/tools/nsc/backend/ScalaPrimitivesOps.scala:10]: If you'll switch names between object that are currently named `ScalaPrimitives` and `ScalaPrimitivesOps` most of the need for api changes will be gone. While this messes up the name, I think it's worth it both for API compatibility and for binary compatibility.

File: src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala [hunk 4]
```diff
@@ -28,8 +28,8 @@ abstract class ConstantOptimization extends SubComponent {
   import global._
   import icodes._
   import icodes.opcodes._
-  import scala.tools.nsc.backend.icode.Primitives._
-  import scala.tools.nsc.backend.icode.Opcodes._
+  import scala.tools.nsc.backend.jvm.Primitives._
```
Comment (DarkDimius) [inline src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:31]: why did you move them? with the same reasoning, I believe we should keep them where they are now.

File: src/compiler/scala/tools/nsc/backend/jvm/GenBCodeOps.scala [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+
+package scala
+package tools.nsc
+package backend
+package jvm
+
+import scala.tools.asm
+
+object GenBCodeOps extends GenBCodeOps
```
Comment (DarkDimius) [inline src/compiler/scala/tools/nsc/backend/jvm/GenBCodeOps.scala:14]: Same with this refactoring.

**Review by DarkDimius — CHANGES_REQUESTED** [review]
See above

File: src/compiler/scala/tools/nsc/backend/ScalaPrimitivesOps.scala (hunk 3 above)
Comment (smarter) [reply src/compiler/scala/tools/nsc/backend/ScalaPrimitivesOps.scala:10]: There is still an `object ScalaPrimitives` for compatibility: https://github.com/DarkDimius/scala/pull/7/files#diff-859cee54c124f77e4b12fe371a1198a6R411, what else is needed?

File: src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala (hunk 4 above)
Comment (smarter) [reply src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:31]: `icode.Primitives` contained both a class and an object, I cannot compile the class with dotty and I need to compile the object, so I had to move the object. I decided to move it to `jvm.Primitives` instead of something like `icode.PrimitivesOps` because that way I don't need to compile anything from the `icode` subdirectory which makes the selection with filters of files compiled by dotty simpler.

File: src/compiler/scala/tools/nsc/backend/jvm/GenBCodeOps.scala (hunk 5 above)
Comment (smarter) [reply src/compiler/scala/tools/nsc/backend/jvm/GenBCodeOps.scala:14]: As in the other cases, I needed to compile the object but not the class, but kept a compatibility `object GenBCode extends GenBCodeOps`

File: src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala (hunk 4 above)
Comment (DarkDimius) [reply src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:31]: Let me think about it more.

Comment (smarter) [reply src/compiler/scala/tools/nsc/backend/opt/ConstantOptimization.scala:31]: I really really would like to avoid that, it's very weird, it may break stuff, it's evil. Also I don't see what difference it would make, you still need to split things into different files with different names.

**Comment by smarter** [comment]
> Let me think about it more.

Any more thoughts on this? Would be good to know how to proceed since this is blocking the hyper bootstrap which is in turn making it harder for Martin to do the changes he wants to do in the backend.

**Comment by DarkDimius** [comment]
After followup discussion we decided that we'll get this in as is.
